ports/nrf/internal_flash.c

```diff
@@ -67,101 +84,63 @@ uint32_t internal_flash_get_block_count(void) {
+    uint32_t src = lba2addr(block);
+    memcpy(dest, (uint8_t*) src, FILESYSTEM_BLOCK_SIZE*num_blocks);
+    return 0; // success
+}
 
-  if (!sd_en) {
-    // Erase
-    NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Een << NVMC_CONFIG_WEN_Pos);
-    while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+mp_uint_t internal_flash_write_blocks(const uint8_t *src, uint32_t lba, uint32_t num_blocks) {
 
-    NRF_NVMC->ERASEPAGE = dest;
-    while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+    #ifdef MICROPY_HW_LED_MSC
+    port_pin_set_output_level(MICROPY_HW_LED_MSC, true);
+    #endif
 
-    NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos);
-    while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+    while (num_blocks) {
+        uint32_t const addr      = lba2addr(lba);
+        uint32_t const page_addr = addr & ~(FL_PAGE_SZ - 1);
 
-    // Write
-    uint32_t *p_src = (uint32_t*) buf;
-    uint32_t *p_dest = flash_align;
-    uint32_t i = 0;
+        uint32_t count = 8 - (lba%8); // up to page boundary
+        count = MIN(num_blocks, count);
 
-    while (i < (FLASH_PAGE_SIZE / sizeof(uint32_t))) {
-      NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos);
-      while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+        if (page_addr != _fl_pg_addr) {
+            internal_flash_flush();
 
-      *p_dest++ = *p_src++;
+            // writing previous cached data, skip current data until flashing is done
+            // tinyusb stack will invoke write_block() with the same parameters later on
+            //        if ( _is_flashing ) return;
 
-      while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+            _fl_pg_addr = page_addr;
+            memcpy(_fl_cache, (void *)page_addr, FL_PAGE_SZ);
+        }
 
-      NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos);
-      while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+        memcpy(_fl_cache + (addr & (FL_PAGE_SZ - 1)), src, count*FILESYSTEM_BLOCK_SIZE);
 
-      ++i;
+        // adjust for next run
+        lba        += count;
+        src        += count*FILESYSTEM_BLOCK_SIZE;
+        num_blocks -= count;
     }
-  }
-
-  clear_temp_status();
-#ifdef MICROPY_HW_LED_MSC
-  port_pin_set_output_level(MICROPY_HW_LED_MSC, false);
-#endif
-  return true;
-}
 
-mp_uint_t internal_flash_read_blocks(uint8_t *dest, uint32_t block, uint32_t num_blocks) {
-  uint32_t src = convert_block_to_flash_addr(block);
-  memcpy(dest, (uint8_t*) src, FILESYSTEM_BLOCK_SIZE*num_blocks);
-  return 0; // success
-}
+    #ifdef MICROPY_HW_LED_MSC
+    port_pin_set_output_level(MICROPY_HW_LED_MSC, false);
+    #endif
 
-mp_uint_t internal_flash_write_blocks(const uint8_t *src, uint32_t block_num, uint32_t num_blocks) {
-  for (size_t i = 0; i < num_blocks; i++) {
-    if (!internal_flash_write_block(src + i * FILESYSTEM_BLOCK_SIZE, block_num + i)) {
-      return 1; // error
-    }
-  }
-  return 0; // success
+    return 0; // success
```

Comment:
No better return value?

Reply:
not entirely sure, since copy it from samd port
https://github.com/adafruit/circuitpython/blob/nrf52840_usbboot/ports/atmel-samd/internal_flash.c#L203